File: Week_06/403/frog_jump.go
package main

type crossParam struct {
	start int
	jump  int
}

func canCross(stones []int) bool {
	stoneSet := make(map[int]bool, 0)
	for _, s := range stones {
		stoneSet[s] = true
	}
	memo := make(map[crossParam]bool, 0)
	var helper func(i, k int) bool
	helper = func(i, k int) bool {
		if i == stones[len(stones)-1] {
			return true
		}
		if v, ok := memo[crossParam{i, k}]; ok {
			return v
		}
		res := false
		for j := k - 1; j <= k+1; j++ {
			if j > 0 && stoneSet[i+j] && helper(i+j, j) {
				res = true
				break
			}
		}
		memo[crossParam{i, k}] = res
		return res
	}
	return helper(0, 0)
}

File: Week_06/403/frog_jump.py
from typing import List


class Solution:
    def canCross(self, stones: List[int]) -> bool:
        if stones[1] != 1:
            return False

        stones_set = set(stones)
        memo = {}

        def dfs(i: int, k: int) -> bool:
            if i == stones[-1]:
                return True
            if (i, k) in memo:
                return memo[(i, k)]
            res = False
            for j in range(k - 1, k + 2):
                if j > 0 and i + j in stones_set and dfs(i + j, j):
                    res = True
                    break
            memo[(i, k)] = res
            return res

        return dfs(0, 0)
